### ISSUE TITLE
StringIO error in ncbi_eutil_esearch

### ISSUE THREAD

**Issue by hepcat72** [galaxyproject/tools-iuc#3111]
I'm not a python expert, and I haven't tried messing with the code (yet), but I think this line:

https://github.com/galaxyproject/tools-iuc/blob/916aaa04992e36f9324299949b86cccd12d3081d/tools/ncbi_entrez_eutils/eutils.py#L91

may be causing this error:

```
Traceback (most recent call last):
  File "/data/galaxy/shed_tools/testtoolshed.g2.bx.psu.edu/repos/iuc/ncbi_eutils_esearch/b88af93de1b7/ncbi_eutils_esearch/esearch.py", line 47, in <module>
    history = c.extract_history(results)
  File "/data/galaxy/shed_tools/testtoolshed.g2.bx.psu.edu/repos/iuc/ncbi_eutils_esearch/b88af93de1b7/ncbi_eutils_esearch/eutils.py", line 91, in extract_history
    parsed_data = Entrez.read(StringIO.StringIO(xml_data))
AttributeError: type object '_io.StringIO' has no attribute 'StringIO'
```

because StringIO is already imported on this line:

https://github.com/galaxyproject/tools-iuc/blob/916aaa04992e36f9324299949b86cccd12d3081d/tools/ncbi_entrez_eutils/eutils.py#L3

...at least, according the second answer here:

https://stackoverflow.com/questions/52440965/attributeerror-type-object-io-stringio-has-no-attribute-stringio

This happens when switching `Store results to history server` to `Yes` and searching the `Gene` database with the query:

`genetype rrna[Properties] AND Homo sapiens[Organism] AND (srcdb refseq[Properties] AND alive[prop])`

Let me know if I did something wrong that's causing this error.

**Comment by peterjc** [comment]
That looks like a Python 2 era reference ``StringIO.StringIO`` (module dot class), should be ``io.StringIO`` in Python 3, or just ``StringIO`` (the class) given how the import was done.

**Comment by hepcat72** [comment]
I just tried reproducing the issue in planemo before trying a fix, but in `planemo serve --galaxy_python_version 3.7`, I get a different error: `ModuleNotFoundError: No module named 'Bio'`.  I tried twice.

**Comment by hepcat72** [comment]
For clarity, I get that error when I try the same search with the esearch tool via planemo.

**Comment by peterjc** [comment]
Missing ``Bio`` means missing Biopython - yet ``ncbi_entrez_eutils/esearch.xml`` has the line ``<expand macro="requirements"/>`` which is a reference to the dependency in ``tools/ncbi_entrez_eutils/macros.xml``:

```xml
  <xml name="requirements">
    <requirements>
      <requirement type="package" version="1.70">biopython</requirement>
    </requirements>
  </xml>
```

**Comment by bernt-matthias** [comment]
@hepcat72 when you run the failing planemo test, do you have a conda environment activated? See this one https://github.com/galaxyproject/planemo/issues/965

**Comment by hepcat72** [comment]
I did.  I was confused when python3 was in my path, yet was getting errors about python 2.7, so I'd tried using a conda environment with python 3.8.3.  I later found out about the `--galaxy_python_version 3.7` flag.  So thanks for this.  I will try without it.

**Comment by hepcat72** [comment]
*sigh* OK.  Without using the conda environment, I can't seem to get `planemo serve` to finish successfully...
```
(.venv) bash-5.0$ planemo serve --galaxy_python_version 3.7
git --git-dir /Users/rleach/.planemo/gx_repo config remote.origin.fetch '+refs/*:refs/*'
git --git-dir /Users/rleach/.planemo/gx_repo config remote.origin.mirror true
git --git-dir /Users/rleach/.planemo/gx_repo remote update >/dev/null 2>&1
galaxy.util.commands WARNING: Passing program arguments as a string may be a security hazard if combined with untrusted input

...

The Galaxy client has not yet been built and will be built now.
Installing node into /Users/rleach/.planemo/gx_venv_3.7 with nodeenv.
./scripts/common_startup.sh: line 240: nodeenv: command not found
Problem running Galaxy command [].
Traceback (most recent call last):
  File "/Users/rleach/.venv/bin/planemo", line 8, in <module>
    sys.exit(planemo())
  File "/Users/rleach/.venv/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/rleach/.venv/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/planemo/cli.py", line 190, in handle_blended_options
    return f(*args, **kwds)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/planemo/commands/cmd_serve.py", line 43, in cli
    galaxy_serve(ctx, runnables, **kwds)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/planemo/galaxy/serve.py", line 22, in serve
    return _serve(ctx, runnables, **kwds)
  File "/Users/rleach/.venv/lib/python3.7/site-packages/planemo/galaxy/serve.py", line 54, in _serve
    raise Exception(message)
Exception: Problem running Galaxy command [].
(.venv) bash-5.0$ 
```
Looking at the progress when I'd used the conda env, `nodeenv` was in that conda env:
```
Requirement already satisfied: nodeenv==1.3.5 in /Users/rleach/.miniconda3/lib/python3.7/site-packages (from -r requirements.txt (line 100)) (1.3.5)
```
Perhaps I should try to add it to the virtualenv?

Here was my planemo installation procedure:

First, I had to use conda to create the virtualenv because I don't have it outside of conda (is this the problem?):
```
. .bash_profile   #This activates conda (because I don't have virtualenv outside of conda)
virtualenv .venv
```
Then I open a new shell without conda activated and:
```
. .venv/bin/activate
pip install "pip>=7"
pip install planemo
```

**Comment by bernt-matthias** [comment]
I think `.venv` is irrelevant since planemo uses `/Users/rleach/.planemo/gx_venv_3.7`.

Maybe remove `/Users/rleach/.planemo/gx_venv_3.7` and try from scratch (make sure to deactivate any conda envs that might be active).

Do you have an alternative to using virtualenv from a conda environment? I remember that you can also create an venv with `python3 -m venv /path/to/new/virtual/environment`. Maybe this can be used to create a new venv `/Users/rleach/.planemo/gx_venv_3.7` which is then filled during the next `planemo test` run.

**Comment by hepcat72** [comment]
That made sense, although even though I blew away all of `~/.planemo`, I still get the same error about `nodeenv`.  I will try installing a system level python3 to eliminate the need for conda when setting up a virtualenv.

```
(.venv) bash-5.0$ history
...
  308  . .bash_profile
  309  virtualenv .venv
  310  exit
  311  . .venv/bin/activate
  312  pip install "pip>=7"
  313  pip install planemo
  314  planemo serve --galaxy_python_version 3.7
  315  history
(.venv) bash-5.0$ which conda
(.venv) bash-5.0$ 
```

**Comment by hepcat72** [comment]
Well, I didn't realize I already had a system level python3 in /usr/bin already, so I tried your second suggestion.  It still has an issue with nodenv...
```
rm -rf .venv .planemo
python3 -m venv /Users/rleach/.venv
. .venv/bin/activate
pip install "pip>=7"
```
The output suggested I upgrade with:
	`pip install --upgrade pip`
which I did, then...
```
pip install planemo
planemo serve --galaxy_python_version 3.7
```

I got a pair of the exception (the same ones) twice, plus a final exception:

```
The Galaxy client has not yet been built and will be built now.
Installing node into /Users/rleach/.planemo/gx_venv_3.7 with nodeenv.
 * Install prebuilt node (10.15.3) .
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 1350, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1240, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1286, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1235, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1006, in _send_output
    self.send(msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 946, in send
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1409, in connect
    self.sock = self._context.wrap_socket(self.sock,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1108)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/rleach/.planemo/gx_venv_3.7/bin/nodeenv", line 8, in <module>
    sys.exit(main())
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 1113, in main
    create_environment(env_dir, opt)
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 936, in create_environment
    install_node(env_dir, src_dir, opt)
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 700, in install_node
    install_node_wrapped(env_dir, src_dir, opt)
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 723, in install_node_wrapped
    download_node_src(node_url, src_dir, opt, prefix)
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 552, in download_node_src
    dl_contents = io.BytesIO(urlopen(node_url).read())
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 580, in urlopen
    return urllib2.urlopen(req)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 542, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 1393, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 1353, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1108)>

...

The Galaxy client has not yet been built and will be built now.
Installing node into /Users/rleach/.planemo/gx_venv_3.7 with nodeenv.
 * Install prebuilt node (10.15.3) .
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 1350, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1240, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1286, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1235, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1006, in _send_output
    self.send(msg)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 946, in send
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py", line 1409, in connect
    self.sock = self._context.wrap_socket(self.sock,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1108)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/rleach/.planemo/gx_venv_3.7/bin/nodeenv", line 8, in <module>
    sys.exit(main())
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 1113, in main
    create_environment(env_dir, opt)
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 936, in create_environment
    install_node(env_dir, src_dir, opt)
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 700, in install_node
    install_node_wrapped(env_dir, src_dir, opt)
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 723, in install_node_wrapped
    download_node_src(node_url, src_dir, opt, prefix)
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 552, in download_node_src
    dl_contents = io.BytesIO(urlopen(node_url).read())
  File "/Users/rleach/.planemo/gx_venv_3.7/lib/python3.8/site-packages/nodeenv.py", line 580, in urlopen
    return urllib2.urlopen(req)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 542, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 1393, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/urllib/request.py", line 1353, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1108)>
Problem running Galaxy command [].
Traceback (most recent call last):
  File "/Users/rleach/.venv/bin/planemo", line 8, in <module>
    sys.exit(planemo())
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/cli.py", line 190, in handle_blended_options
    return f(*args, **kwds)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/commands/cmd_serve.py", line 43, in cli
    galaxy_serve(ctx, runnables, **kwds)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/galaxy/serve.py", line 22, in serve
    return _serve(ctx, runnables, **kwds)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/galaxy/serve.py", line 54, in _serve
    raise Exception(message)
Exception: Problem running Galaxy command [].
```
It appears to be an SSL certificate issue?

**Comment by bernt-matthias** [comment]
Hmm. How about adding `--galaxy_branch release_20.05`. Btw is `planemo test` working?

**Comment by bernt-matthias** [comment]
Maybe https://github.com/galaxyproject/training-material/pull/1784/files helps. Or other issues found here https://github.com/search?q=org%3Agalaxyproject+CERTIFICATE_VERIFY_FAILED&type=Issues

**Comment by hepcat72** [comment]
`planemo test` fails.  I'm not sure where the path on that last line is coming from (in the DistutilsFileError error).  How could that path be getting in there?

```
(.venv) bash-5.0$ planemo test
Traceback (most recent call last):
  File "/Users/rleach/.venv/bin/planemo", line 8, in <module>
    sys.exit(planemo())
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/cli.py", line 190, in handle_blended_options
    return f(*args, **kwds)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/commands/cmd_test.py", line 80, in cli
    runnables = for_paths(paths, temp_path=temp_path)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/runnable.py", line 149, in for_paths
    return [for_path(path, temp_path=temp_path) for path in paths]
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/runnable.py", line 149, in <listcomp>
    return [for_path(path, temp_path=temp_path) for path in paths]
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/runnable.py", line 142, in for_path
    path = _copy_runnable_tree(path, runnable_type, temp_path)
  File "/Users/rleach/.venv/lib/python3.8/site-packages/planemo/runnable.py", line 115, in _copy_runnable_tree
    copy_tree(dir_to_copy, temp_path)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/dir_util.py", line 157, in copy_tree
    copy_tree(src_name, dst_name, preserve_mode,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/dir_util.py", line 157, in copy_tree
    copy_tree(src_name, dst_name, preserve_mode,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/dir_util.py", line 157, in copy_tree
    copy_tree(src_name, dst_name, preserve_mode,
  [Previous line repeated 1 more time]
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/dir_util.py", line 161, in copy_tree
    copy_file(src_name, dst_name, preserve_mode,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/distutils/file_util.py", line 104, in copy_file
    raise DistutilsFileError(
distutils.errors.DistutilsFileError: can't copy '/Users/rleach/GoogleDrivePU/PROJECT/REPOS/htseq/htseq': doesn't exist or not a regular file
(.venv) bash-5.0$ pwd
/Users/rleach
```

I will try `--galaxy_branch release_20.05`...

**Comment by hepcat72** [comment]
Ugh.  Same exception when adding `--galaxy_branch release_20.05` to the planemo command.  Lance just suggested I run `pip install --upgrade certifi`.  Trying that now...

**Comment by hepcat72** [comment]
`pip install --upgrade certifi` says:
```
Requirement already up-to-date: certifi in ./.venv/lib/python3.8/site-packages (2020.6.20)
```

**Comment by hepcat72** [comment]
It appears that double-clicking this file fixes the cert issue:
```
/Applications/Python 3.8/Install Certificates.command
```
At least, it seems to be using nodeenv now.  It's still running...

**Comment by hepcat72** [comment]
Yay!  `planemo serve` now works!  Now back to the original issue...

**Comment by hepcat72** [comment]
OK, as the stack thread suggests, changing:

`parsed_data = Entrez.read(StringIO.StringIO(xml_data))`

to:

`parsed_data = Entrez.read(StringIO(xml_data))`

fixes the issue.  PR impending.

**Comment by hepcat72** [comment]
So given this is a python 2 versus 3 issue, is there anything I should do to not break 2, or should I not bother with backward compatibility?

**Comment by bernt-matthias** [comment]
If I'm not wrong the `io` module is py3 specific anyway. So go ahead please :)

**Comment by bernt-matthias** [comment]
Maybe you can also have a look at the PR https://github.com/galaxyproject/tools-iuc/pull/791 .. in particular the issue raised here https://github.com/galaxyproject/tools-iuc/pull/791#issuecomment-225224370

**Comment by hepcat72** [comment]
Yes actually, I had noticed the same thing and was composing a similar issue regarding the xml/json issue (among other issues like not being able to generate a list of IDs from the search that could be provided to efetch's ID list input, or being able to view the results (in 20.01 - fixed in 20.05)), but before submitting, I tried testing and ran into this issue.  I have a number of observations about the tools that I would like to look into.

**Comment by hepcat72** [comment]
OK. I just read through the other issue after figuring things out a bit WRT the galaxy tool wrapper for Entrez eutils. I have confirmed that Biopython's `Entrez.efetch()` method will never work when supplied a history as a source for IDs to retrieve records for.  The eutils NCBI API can utilize a history as the ID source, but the biopython code never attempts to use it.  It only takes a list of IDs (either by file or as a string).  I wrote a quick one-off version of the galaxy wrapper's efetch.py which circumvents the biopython efetch method and directly calls it's private `_open()` method (supplying it the webenv string and query_key), and it works.  I'm manually able to supply a `webenv` and `query_key` value and retrieve fasta records.

I also updated `elink.{py,xml}` in the galaxy wrapper to make it work and added `linkname` and `retmode` parameters so I could better control the types of IDs I got back and could change the type to json.  However, after adding select lists for all the link table options, the tool form takes maybe 10 seconds to process (which is quite a long time) and I have not updated the parsing of the return xml or json to retrieve the correct query key or webenv string needed to pass to efetch.  (Perhaps there's a way to get it to return just that?)  Also, it currently can create a history file and a file of IDs, but they're the same file.

I could finish this one part to be able to accomplish my immediate goal, but changes would be needed in biopython as well, and it appears that these things are just the tip of the iceberg.

**Comment by peterjc** [comment]
If you've run into a bug with the Biopython Entrez efetch with history, please open an issue here: https://github.com/biopython/biopython/issues

There are examples using it to combine esearch/efetch in our documentation, http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec161

**Comment by hepcat72** [comment]
The issue appears to be with elink/efetch. You cannot provide a history from elink to it.

**Comment by hepcat72** [comment]
Just a further clarification/correction...

Yes, esearch->efetch via history works, but if you search (e.g.) the "gene" db, and want the linked "nuccore" fasta records associated with the gene db's UIDs that are returned, you have to go through elink (I thought it was notable that if you enter N gene UIDs and select "nuccore" in efetch, instead of getting no results, you get what appear to be the first N records from nuccore in some sort of alphanumeric order - which is perhaps understandable, but a bit confusing for an inexperienced user).

Going through elink, you can get either json or xml (it does not currently return a file of UIDs).  If you provide that elink history output to efetch, in either case, you get a red/empty collection.  The json(/xml?) structure output by elink puts webenv and querykey in a place that the tools-iuc eutils suite does not currently parse.

The only way in galaxy currently to provide elink output to efetch is the history (though perhaps it can be changed to output an ID list file?).  The efetch galaxy tool allows you to select the elink history output when you change the output to json(/xml?), but since none of the IDs, webenv, or query_key items are successfully parsed out by efetch.py, it cannot supply those items in the payload and the red/empty collection is the result.

I was incorrect in my initial assessment of the biopython Entrez package's efetch code (python's not my native language). It will succeed if you pass it a value for webenv and query_key (and no `id` value) in the payload.  So the issue is just obtaining those values from the elink output and putting them in the payload supplied to Entrez.efetch.

Since webenv strings expire, it might be a good idea to at least convert the elink output into a ID list file and offer the history (json/xml) as an optional output.

**Comment by peterjc** [comment]
Thanks for the clarification.

You're right, the webenv and key strings are transient, which has implications for partially re-running a workflow. It should be OK if you re-run an entire workflow as that would get a fresh webenv/key.

**Comment by hepcat72** [comment]
For my testing, I've been using a particular search/link/fetch, and I'd noticed that I needed to change the rettype and/or retmode in the macros for efetch to return fasta output. So I guessed maybe these specs had changed.  I was looking at:

https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/?report=objectonly

and I noticed that efetch supports only 20 of the 38 NLM databases.  But I also noticed that that table doesn't mention json at all, which I know works.  Is there another source I should be using to update the retmodes and rettypes?  I figure maybe I should create a separate select list for efetch since it doesn't support all the databases too...
